### PR TITLE
Add tab switcher

### LINKED ISSUE #7653: Tab Switcher
### Check for existing issues

- [X] Completed

### Describe the feature

When working on tabless editing mode (#6424), [there was a discussion](https://github.com/zed-industries/zed/pull/2742#issuecomment-1638903687) about the need to implement Tab Switcher -- a panel similar to the File Finder that would show the list of opened tabs in the active pane and allow to switch between them and close them right in this panel.

I tried to figure out how such switcher could look and how it could be implemented, and decided to file it as a separate feature request, because there's a lot of things to discuss (and probably to change in the code).

## How it will look and work

zed-tab-switcher

- Default shortcut is `ctrl-tab`. It also moves selection to the next item.
- Selection is confirmed by either releasing of a modifier key (`ctrl` by default) or with a mouse click.
- Items are sorted in the reverse order of the tab activation history.
- Second item (if exists) is selected by default. This will allow to switch to the previous tab with a single keystroke.
- We need item icons here to distinguish files from terminal, search results etc.
- Close button is displayed when hover over item.
- Implemented using the existing Picker functionality, but with the query editor hidden.

## Tasks

- [x] Implement #8757
- [x] Implement headless picker #9099
- [ ] Implement tab switcher

## How others do this

See comments below:
- [VSCode](https://github.com/zed-industries/zed/issues/7653#issuecomment-1936952462)
- [IntelliJ IDEA](https://github.com/zed-industries/zed/issues/7653#issuecomment-1936952779)
- [Sublime Text](https://github.com/zed-industries/zed/issues/7653#issuecomment-1936952882)

## Related issues and PRs

- #6424
- https://github.com/zed-industries/zed/pull/2742#issuecomment-1638903687
- #7356
- #8757

### PR DESCRIPTION
**Pull request by alygin**
The Tab Switcher implementation (#7653):
- `ctrl-tab` opens the Tab Switcher and moves selection to the previously selcted tab. It also cycles selection forward.
- `ctrl-shift-tab` opens the Tab Switcher and moves selection to the last tab in the list. It also cycles selection backward.
- Tab is selected and the Tab Switcher is closed on the shortcut modifier key (`ctrl` by default) release.
- List items are in reverse activation history order.
- The list reacts to the item changes in background (new tab, tab closed, tab title changed etc.)

Intentionally not in scope of this PR:
- File icons
- Close buttons

I will come back to these features. I think they need to be implemented in separate PRs, and be synchronized with changes in how tabs are rendered, to reuse the code as it's done in the current implementation. The Tab Switcher looks usable even without them.

Related issues:

- Unblocks #7356, I hope 😄

How it looks and works (it's only `ctrl-tab`'s and `ctrl-shift-tab`'s, no `enter`'s or mouse clicks):

https://github.com/zed-industries/zed/assets/2101250/4ad4ec6a-5314-481b-8b35-7ac85e43eb92

fixes #7653, #7321

Release Notes:

- Added Tab Switcher which is accessible via `ctrl-tab` and `ctrl-shift-tab` (#7653) (#7321)

### CODE REVIEW

**Comment by RemcoSmitsDev** [comment]
I really like the feature! But I'm missing a default keybind configuration for it.

**Comment by alygin** [comment]
> I really like the feature! But I'm missing a default keybind configuration for it.

What keybinding do you think should be the default one for the tab switchier?

I also prefer having a shortcut to this feature, but none of the editors/IDEs I use provide it. VSCode doesn't even allow to set a custom one. I guess, the vast majority of the users prefer just clicking tabs, or navigate through them in some other manner. So I'm hesitant to propose a default keybinding, given that the user can always set the one he or she prefers.

**Comment by fostertheweb** [comment]
Ctrl+Tab in VS Code gives you something similar so that may be a reasonable keyboard shortcut. That is also common behavior in web browsers to page through your tabs with Ctrl+Tab. An alternative that comes to mind is Cmd+B because I think of tabs as buffers as a vim user.

**Comment by alygin** [comment]
@fostertheweb, you're right, the `ctrl+tab` in VSCode is a good example of how the tab switcher could look and behave. In the JetBrains products it works the same way, but provides more functionality. Unfortunately, the current state of the shortcuts handling in Zed doesn't allow to implement the such behaviour easily ([the same problem with the `cmd+p` in the file finder](https://github.com/zed-industries/zed/issues/4450#issuecomment-1908621350)), but I'll give it another look.

`cmd+b` is already taken for the left dock show/hide, though `cmd-k cmd-b` is still vacant :)

**Comment by d1y** [comment]
Regarding the key bindings, it’s just my personal opinion:
- shift-shift: JetBrains
- shift-`: More affectionate

**Comment by alygin** [comment]
> I wonder how much more work it would be to allow for Ctrl+Tab.

Here's a discussion of this feature: [Action Release trigger](https://github.com/zed-industries/zed/discussions/8310)

File: crates/gpui/src/platform/keystroke.rs [hunk 1]
```diff
@@ -213,6 +213,14 @@ impl Modifiers {
         }
     }
 
+    /// helper method for Modifiers with just control
+    pub fn control() -> Modifiers {
+        Modifiers {
+            command: true,
```
Comment (ConradIrwin) [inline crates/gpui/src/platform/keystroke.rs:219]:
```suggestion
            control: true,
```

Comment (alygin) [reply crates/gpui/src/platform/keystroke.rs:219]: 🤦🏻‍♂️

**Review by mikayla-maki — APPROVED** [review]
This looks great, thank you for the extensive tests!

**Comment by thynus** [comment]
Well I discovered that CTRL-` will cycle between the shell and the upper pane so that's what I was looking for.

**Comment by samjeffress** [comment]
I found it confusing that the keymap for the tab switcher was only in the Workspace context. I added the bindings to the Editor context too as that's where I usually am moving betwween files. Did I miss something obvious?

**Comment by alygin** [comment]
@samjeffress, workspace contains editors, so whatever binding you add to the workspace context should work in the editor context too. Doesn't `ctrl+tab` work for you in an editor whey you use the default configuration?

**Comment by samjeffress** [comment]
@alygin with no custom bindings for "ctrl-tab", pressing "ctrl-tab" toggles between tabs without showing the list of open tabs (when in the editor). When I add the following to keymap.json the tab switcher appears:
```
  {
    "context": "Editor",
    "bindings": {
      "ctrl-tab": "tab_switcher::Toggle",
      "ctrl-shift-tab": [
        "tab_switcher::Toggle",
        {
          "select_last": true
        }
      ]
    }
  },
```

**Comment by anuragg-p** [comment]
Hi @samjeffress , 
Any idea on how to do this better with vim mode on?
Just like we do in telescope?

**Pull request merged**